File: system_monitor/disk_discovery.py
from __future__ import annotations

import re

import psutil

from .config_loader import SensorConfig

_IGNORE_FSTYPES = frozenset(
    {
        "tmpfs",
        "devtmpfs",
        "squashfs",
        "overlay",
        "proc",
        "sysfs",
        "devfs",
        "autofs",
        "cgroup",
        "cgroup2",
        "pstore",
        "bpf",
        "tracefs",
        "debugfs",
        "securityfs",
        "configfs",
        "fusectl",
        "mqueue",
        "hugetlbfs",
    }
)


def _mount_to_sensor_id(mountpoint: str) -> str:
    cleaned = mountpoint.strip().rstrip("\\/").lower()
    if not cleaned:
        return "disk_auto_root"
    cleaned = cleaned.replace(":", "").replace("\\", "_").replace("/", "_")
    cleaned = re.sub(r"[^a-z0-9_]+", "_", cleaned)
    cleaned = re.sub(r"_+", "_", cleaned).strip("_")
    return f"disk_auto_{cleaned or 'root'}"


def discover_disk_sensors() -> list[SensorConfig]:
    import platform

    sensors: list[SensorConfig] = []
    seen_ids: set[str] = set()
    all_partitions = platform.system() != "Windows"

    for part in psutil.disk_partitions(all=all_partitions):
        if part.fstype in _IGNORE_FSTYPES:
            continue
        try:
            psutil.disk_usage(part.mountpoint)
        except (OSError, PermissionError):
            continue

        sensor_id = _mount_to_sensor_id(part.mountpoint)
        if sensor_id in seen_ids:
            suffix = 2
            while f"{sensor_id}_{suffix}" in seen_ids:
                suffix += 1
            sensor_id = f"{sensor_id}_{suffix}"
        seen_ids.add(sensor_id)

        label = part.mountpoint
        if part.device:
            label = f"{part.mountpoint} ({part.device})"

        sensors.append(
            SensorConfig(
                id=sensor_id,
                name=f"Диск {label}",
                type="system.disk_usage",
                enabled=True,
                interval_sec=30,
                unit="%",
                params={"path": part.mountpoint},
                warn_above=85,
                critical_above=95,
            )
        )

    return sensors

File: internal/diskdiscovery/diskdiscovery.go
package diskdiscovery

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/pagbest154-cmd/system-monitor/internal/config"
	"github.com/shirou/gopsutil/v4/disk"
)

var ignoreFSTypes = map[string]bool{
	"tmpfs": true, "devtmpfs": true, "squashfs": true, "overlay": true,
	"proc": true, "sysfs": true, "devfs": true, "autofs": true,
	"cgroup": true, "cgroup2": true, "pstore": true, "bpf": true,
	"tracefs": true, "debugfs": true, "securityfs": true, "configfs": true,
	"fusectl": true, "mqueue": true, "hugetlbfs": true,
}

func MountToSensorID(mountpoint string) string {
	cleaned := strings.Trim(strings.TrimRight(mountpoint, "\\/"), " ")
	if cleaned == "" {
		return "disk_auto_root"
	}
	cleaned = strings.ReplaceAll(cleaned, ":", "")
	cleaned = strings.ReplaceAll(cleaned, "\\", "_")
	cleaned = strings.ReplaceAll(cleaned, "/", "_")
	re := regexp.MustCompile(`[^a-z0-9_]+`)
	cleaned = strings.ToLower(cleaned)
	cleaned = re.ReplaceAllString(cleaned, "_")
	re2 := regexp.MustCompile(`_+`)
	cleaned = strings.Trim(re2.ReplaceAllString(cleaned, "_"), "_")
	if cleaned == "" {
		cleaned = "root"
	}
	return "disk_auto_" + cleaned
}

func DiscoverDiskSensors() []config.SensorConfig {
	partitions, err := disk.Partitions(true)
	if err != nil {
		partitions, _ = disk.Partitions(false)
	}
	sensors := make([]config.SensorConfig, 0)
	seen := map[string]bool{}
	for _, part := range partitions {
		if ignoreFSTypes[part.Fstype] {
			continue
		}
		if _, err := disk.Usage(part.Mountpoint); err != nil {
			continue
		}
		sensorID := MountToSensorID(part.Mountpoint)
		if seen[sensorID] {
			suffix := 2
			for seen[sensorID+"_"+strconv.Itoa(suffix)] {
				suffix++
			}
			sensorID = sensorID + "_" + strconv.Itoa(suffix)
		}
		seen[sensorID] = true
		label := part.Mountpoint
		if part.Device != "" {
			label = part.Mountpoint + " (" + part.Device + ")"
		}
		warn := 85.0
		critical := 95.0
		interval := 30
		sensors = append(sensors, config.SensorConfig{
			ID:            sensorID,
			Name:          "Диск " + label,
			Type:          "system.disk_usage",
			Enabled:       true,
			IntervalSec:   &interval,
			Unit:          "%",
			Params:        map[string]interface{}{"path": part.Mountpoint},
			WarnAbove:     &warn,
			CriticalAbove: &critical,
		})
	}
	return sensors
}
